### ISSUE TITLE
Feature Request: SSL Support

### ISSUE THREAD

**Issue by kellymears** [roots/bud#934]
### Summary

Make it as simple as possible to develop using the `https:` protocol in @roots/bud projects. 

See: https://github.com/roots/bud/pulls/933

#### Why are we doing this?

It's increasingly important to develop with SSL enabled

#### What use cases does it support?

- dev parity
- interoperability with other tools

#### What is the expected outcome?

A path forward for configuring Express with user supplied certs that is:
- viable for a general audience
- documented

#### Potential conflicts / foreseeable issues

I don't want bud to be in the business of managing and issuing certificates.

**Comment by retlehs** [comment]
imo this should be critical/one of the highest priority todos

**Comment by strarsis** [comment]
+1! Currently I can't run bud dev/watch with a HTTPS localhost dev server.
Somewhere inside https://github.com/roots/bud/tree/d3a6070118dfd27783b6bdabd75cbd3b9cf7a2bb/packages/%40roots/bud-server/src/middleware/proxy ...

**Comment by kellymears** [comment]
@strasis there is a filter that is relevant here if you have time or willingness to tool around and try to find a configuration that works for you. If you land it, let me know (or submit a PR) to change the provided defaults.

Those options are filtered here (as of 5.2.0):

https://github.com/roots/bud/blob/0f1eadb73f18309107b9fb741fe899ab1d8dc783/sources/%40roots/bud-server/src/middleware/proxy/proxy.middleware.ts#L26-L52

You can tap this filter like:

```ts
app.hooks.on('proxy.options', options => ({
  ...options,
  // your overrides
})
```

this is the middleware https://github.com/chimurai/http-proxy-middleware

**Comment by strarsis** [comment]
@kellymears, @retlehs: Alright, I got HTTPS/SSL support working. `bud` server can be connected to via HTTPS (using a snakeoil* CA cert by default, that can be confirmed, but also an own CA cert is possible (as I am using for development)).

*snakeoil CA certs are used because HTTPS requires some key + cert, even invalid ones, otherwise the SSL handshake won't work. Browsers and other clients usually allow skipping the invalid certificate error. For avoiding those errors and for improved security and business environments, an own CA key + cert can be used, which the development systems and devices already trust (added to trust store).

But just changing the `proxy.options` options object for the bud server is not enough. Some code also needs to change.

`VSCode` + `nodejs` debugging proved to be were very helpful by the way!
And luckily the JavaScript compiled from TypeScript is still nicely readable and understandable.
This allowed for some quick tinkering.

So the `bud` server service uses `express` (it is an `express` app), with 
`https` and `http` middleware for proxying requests to it to the development server:
https://github.com/roots/bud/blob/97c0297bd39b1b2893da83966ce30976b07a1b92/sources/%40roots/bud-server/src/server/index.ts#L78

It doesn't matter how the proxy part is configured as it is the client for the development server and connecting to HTTPS already works.

The goal is to make the `bud` server, the `express` app, listen with HTTPS protocol support.
This can be achieved by making an `express` app HTTPS aware by attaching it to a `nodejs` `https` server instance:
Again, in [`bud-server/src/server/index.ts`](https://github.com/roots/bud/blob/97c0297bd39b1b2893da83966ce30976b07a1b92/sources/%40roots/bud-server/src/server/index.ts):
```js
[...]
const server_watcher_1 = require("./server.watcher");
const https = require('https');
const snakeoil = require("@dazld/snakeoil-certs");

[...]

        /**
         * Listen
         */
        this.httpsServer = https.createServer({
            key:  snakeoil.serviceKey,  //fs.readFileSync('./dev/ssl/ssl-cert-snakeoil.key'),
            cert: snakeoil.certificate, //fs.readFileSync('./dev/ssl/ssl-cert-snakeoil.pem'),
        }, this.application)
        .listen(this.port);

        //this.instance = this.application.listen(this.port, async (error) => {
        this.httpsServer.on('listening', () => {
            this.log('info', `started server on %s`, this.port);
            this.app.hooks.filter('event.server.listen');
        });

        this.httpsServer.on('error', (error) => {
          this.log('error', error);
        });
````

Now HTTPS connections to bud server are already possible - but no plain HTTP connections (plain HTTP connections can be re-added, too, by also attaching a `http` instance, but we are for now interested in HTTPS support, so we leave this out, this also helps with finding out if still plain HTTP connections are somewhere attempted).
When viewing a site using `bud` server one will notice that assets won't load as the URLs are still rewritten with the plain `http` protocol.
And this can be fixed by changing the options object:
```bud.config.js
module.exports = async (app) =>
  app

  // proxy HTTPS config
  .hooks.on('proxy.options', (makeOptions) => {
    let opts = makeOptions() // options _factory_ is passed in


    opts.protocolRewrite = 'https'

    return opts
  })

````
By the way, the `proxy.options` hook handlers will get an options factory function passed, not the actual options object - 
and have to call the options factory function first to get the object for manipulating and returning.

So this is basically all what is needed for HTTPS/TLS support with `bud` server.
It would be a good idea to add this directly to `bud` server and maybe expose a HTTPS toggle option and 
options for paths to a custom CA certificate and key (as I am already using it for development).

Edit: Hm, everything works, but one thing I hadn't gotten perfect, in my test the rewritten URLs don't contain the proxy port. I have to adjust the rewrite function a bit. But HTTPS works.

**Comment by kellymears** [comment]
this is pretty similar to an earlier i attempt i made! Thank you @strarsis 

The listeners:

https://github.com/roots/bud/blob/b679775f3369da31f6696e7945e21657d4283e45/packages/%40roots/bud-server/src/server/index.ts#L172-L216

http server:

https://github.com/roots/bud/blob/b679775f3369da31f6696e7945e21657d4283e45/packages/%40roots/bud-server/src/server/server.http.ts#L4-L36

https server:

https://github.com/roots/bud/blob/b679775f3369da31f6696e7945e21657d4283e45/packages/%40roots/bud-server/src/server/server.https.ts#L5-L60

I'll try to extract these and follow along with the above documentation and see if we can't get this wrapped.

**Comment by kellymears** [comment]
I've got this largely done. 

there are a slew of new hooks in the proxy middleware factory to tap into:

https://github.com/roots/bud/blob/e9ef58110f0f7b26b449ebb52c811731404dfa8c/sources/%40roots/bud-server/src/middleware/proxy/index.ts#L30-L128

an https server:

https://github.com/roots/bud/blob/improve-server/sources/%40roots/bud-server/src/server/server.https.ts#L18-L92

https server will get used alongside http if `dev.ssl.enabled` is `true` and `dev.ssl.cert` + `dev.ssl.key` both have a value set:

https://github.com/roots/bud/blob/e9ef58110f0f7b26b449ebb52c811731404dfa8c/sources/%40roots/bud-server/src/server/index.ts#L150-L166

this set of conditions will also enable the `protocolRewrite` option @strasis pointed out:

https://github.com/roots/bud/blob/e9ef58110f0f7b26b449ebb52c811731404dfa8c/sources/%40roots/bud-server/src/middleware/proxy/index.ts#L102-L105

my only concern is with issuing the fake certs -- i don't have a problem with it conceptually but @dazld/snakeoil-certs hasn't been published to in 6 years. 

i think we either gotta find a different package or document it:

```sh
openssl req -nodes -new -x509 -keyout server.key -out server.cert
```

**Comment by strarsis** [comment]
@kellymears: Concerning the custom cert, the `step` CLI tool (from `smallstep`) can be used - it can also add it to the local trust stores: https://smallstep.com/docs/step-cli/reference/certificate/create

As a default, some cert has to be used, otherwise a protocol error occurs. The snakeoil certs are as untrustworthy as they can get. The package hadn't been updated, though those certs don't need updating either, they had been invalid from the start.

**Comment by kellymears** [comment]
@strarsis makes sense. step seems really cool. i've never seen it before.

i have some tests to fix in #1069 and then i'll see if i can get this working with the new hooks:

```ts
app.hooks.on('dev.ssl.enabled', true)
app.hooks.on('dev.ssl.key', './path/to/dev.key')
app.hooks.on('dev.ssl.cert', './path/to/dev.cert')
app.hooks.on('dev.url', new URL('https://example.test:3000'))
```

if that works then it's just the matter of passing those options through with the `app.serve` facade and we should be good to go. 

Thanks again for your help getting this rolling!

cc @retlehs @swalkinshaw

**Comment by kellymears** [comment]
Merged #1069 🎉. Going to go ahead and close this feature request. If there are issues with my implementation we can address in new issues.

I'll make sure to give @strarsis credit in the release ✌🏽

**Comment by retlehs** [comment]
🎉 🎉 🎉 
💜 💜 💜
